core/contractsdk/cpp/example/charity_demo.cc

```diff
@@ -0,0 +1,418 @@
+     */
+    virtual void queryDonates() = 0;
+    /*
+     * func: 查询拨款记录
+     * @param: startid: 起始记录id
+     * @param: limit: 查询多少条(每次查询不超过100条)
+     */
+    virtual void queryCosts() = 0;
+};
+
+struct CharityDemo : public Charity, public xchain::Contract {
+private:
+    const std::string USERDONATE = "UserDonate_";
+    const std::string ALLDONATE = "AllDonate_";
+    const std::string ALLCOST = "AllCost_";
+    const std::string TOTALRECEIVED = "TotalDonates";
+    const std::string TOTALCOSTS = "TotalCosts";
+    const std::string BALANCE = "Balance";
+    const std::string DONATECOUNT = "DonateCount";
+    const std::string COSTCOUNT = "CostCount";
+    const std::string ADMIN = "admin";
+    const uint64_t MAX_LIMIT = 100;
+
+    std::string getIDFromNum(uint64_t num) {
+        std::ostringstream ss;
+        ss << std::setw(20) << std::setfill('0') << num;
+        return ss.str();
+    }
+
+    bool safe_stoull(const std::string in, uint64_t* out) {
+        if (in.empty()) {
+            return false;
+        }
+        for (int i = 0; i < in.size(); i++) {
+            if (in[i] < '0' || in[i] > '9') {
+                return false;
+            }
+        }
+        std::string::size_type sz = 0;
+        *out = std::stoull(in, &sz);
+        if (sz != in.size()) {
+            return false;
+        }
+        return true;
+    }
+
+    bool isAdmin(xchain::Context* ctx, const std::string& caller) {
+        std::string admin;
+        if (!ctx->get_object(ADMIN, &admin)) {
+            return false;
+        }
+        return (admin == caller);
+    }
+
+public:
+    void initialize() {
+        xchain::Context* ctx = this->context();
+        const std::string& admin = ctx->arg(ADMIN);
+        if (admin.empty()) {
+            ctx->error("missing admin address");
+            return;
+        }
+        ctx->put_object(ADMIN, admin);
+        ctx->put_object(TOTALRECEIVED, "0");
+        ctx->put_object(TOTALCOSTS, "0");
+        ctx->put_object(BALANCE, "0");
+        ctx->put_object(DONATECOUNT, "0");
+        ctx->put_object(COSTCOUNT, "0");
+        ctx->ok("initialize success");
+    }
+
+    void donate() {
+        xchain::Context* ctx = this->context();
+        const std::string& caller = ctx->initiator();
+        if (caller.empty()) {
+            ctx->error("missing initiator");
+            return;
+        }
+
+        if (!isAdmin(ctx, caller)) {
+            ctx->error("only the admin can add donate record");
+            return;
+        }
+
+        const std::string& donor = ctx->arg("donor");
+        if (donor.empty()) {
+            ctx->error("missing 'donor'");
+            return;
+        }
+
+        const std::string& amountStr = ctx->arg("amount");
+        if (amountStr.empty()) {
+            ctx->error("missing 'amount'");
+            return;
+        }
+
+        uint64_t amount;
+        if (!safe_stoull(amountStr, &amount)) {
+            ctx->error(
```

Comment:
why two lines?

Reply:
to keep each line shorter than 80 characters.